python/GafferSceneUI/MeshNormalsUI.py

```diff
@@ -0,0 +1,114 @@
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import Gaffer
+import GafferScene
+import IECoreScene
+
+
+Gaffer.Metadata.registerNode(
+
+	GafferScene.MeshNormals,
+
+	"description",
+	"""
+	Create a normal primitive variable on a mesh, using the positions of adjacent vertices.
+	""",
+
+	"layout:activator:weighting", lambda parent : parent["interpolation"].getValue() != int(IECoreScene.PrimitiveVariable.Interpolation.Uniform),
+	"layout:activator:thresholdAngle", lambda parent : parent["interpolation"].getValue() == int(IECoreScene.PrimitiveVariable.Interpolation.FaceVarying),
+
+	plugs = {
+
+		"interpolation" : [
+			"description",
+			"""
+			The interpolation of the normal primitive variable we are creating. Affects the shape of the resulting normals, because Uniform ( Per-Face ) normals are inherently faceted, whereas Vertex normals are always smooth.
+			""",
+
+			"preset:Uniform (Faceted)", IECoreScene.PrimitiveVariable.Interpolation.Uniform,
+			"preset:Vertex (Smooth)", IECoreScene.PrimitiveVariable.Interpolation.Vertex,
+			"preset:FaceVarying (Depends on threshold)", IECoreScene.PrimitiveVariable.Interpolation.FaceVarying,
+
+			"plugValueWidget:type", "GafferUI.PresetsPlugValueWidget",
+		],
+
+		"weighting" : [
+			"description",
+			"""
+			How to weight the multiple faces that contribute to the normal of a vertex.
+			"Equal" average all faces connected to the vertex - simple to compute, but low quality.
+			"Weight by Angle" gives good results for most meshes.
+			"Weight by Area" may give good results on hard edge models with tight chamfers and large flat faces.
+			""",
+
+			"preset:Equal", IECoreScene.MeshAlgo.NormalWeighting.Equal,
+			"preset:Weight by Angle", IECoreScene.MeshAlgo.NormalWeighting.Angle,
+			"preset:Weight by Area", IECoreScene.MeshAlgo.NormalWeighting.Area,
+
+			"plugValueWidget:type", "GafferUI.PresetsPlugValueWidget",
+
+			"layout:activator", "weighting",
+		],
+
+		"thresholdAngle" : [
+			"description",
+			"""
+			Used to decide whether edges are smooth or sharp when generating a normal primvar with FaceVarying
+			interpolation. FaceVertices with normals that differ by less than this angle will be averaged
```

Comment:
Is this the usual definition for a threshold angle? Whenever I play with it I seem to intuitively expect the opposite - stuff _beyond_ a certain angle _not_ getting averaged.

Reply:
I'm a bit confused by John's description of expecting "stuff beyond a certain angle not getting averaged", which seems like a synonym for "stuff less than an angle getting averaged".

The reasonable other way of thinking of this that I considered was to specify a total angle subtended by the corner, rather  than a difference of normals. So a flat plane would be considered all 180 degree corners, instead of being considered all 0 degree corners, and the threshold would work the opposite way - maybe this is what John was expecting? But as Murray says, common software all seems to work based on "difference of normals".